src/v2/manifest/mod.rs

```diff
@@ -19,62 +19,68 @@ impl Client {
+    }
+
+    /// Fetch an image manifest and return it with its digest.
+    ///
+    /// The name and reference parameters identify the image.
+    /// The reference may be either a tag or digest.
+    pub fn get_manifest_and_ref(&self, name: &str, reference: &str) -> FutureManifestAndRef {
+        let url = {
+            let ep = format!(
+                "{}/v2/{}/manifests/{}",
+                self.base_url.clone(),
+                name,
+                reference
+            );
+            match reqwest::Url::parse(&ep) {
+                Ok(url) => url,
                 Err(e) => {
-                    let msg = format!("new_request failed: {}", e);
-                    error!("{}", msg);
-                    return Box::new(futures::future::err(Error::from(msg)));
+                    return Box::new(futures::future::err::<_, _>(Error::from(format!(
+                        "failed to parse url from string '{}': {}",
+                        ep, e
+                    ))));
                 }
-            };
-            let mtype = mediatypes::MediaTypes::ManifestV2S2.to_string();
-            req.headers_mut().append(
-                header::ACCEPT,
-                match header::HeaderValue::from_str(&mtype) {
-                    Ok(headervalue) => headervalue,
-                    Err(e) => {
-                        let msg = format!("failed to parse HeaderValue from str: {}:", e);
-                        error!("{}", msg);
-                        return Box::new(futures::future::err::<_, _>(Error::from(msg)));
-                    }
-                },
-            );
-            req
+            }
         };
-        let freq = self.hclient.request(req);
-        let fres = freq
-            .from_err()
-            .inspect(move |_| {
-                trace!("GET {:?}", url);
-            })
-            .and_then(|r| {
-                let status = r.status();
+
+        let fres = self
+            .build_reqwest(reqwest::async::Client::new().get(url))
+            .send()
+            .map_err(|e| ::errors::Error::from(format!("{}", e)))
+            .and_then(|res| {
+                trace!("GET '{}' status: {:?}", res.url(), res.status());
+                let status = res.status();
                 trace!("Got status: {:?}", status);
                 match status {
-                    hyper::StatusCode::OK => Ok(r),
-                    _ => Err(format!("get_manifest: wrong HTTP status '{}'", status).into()),
+                    reqwest::StatusCode::OK => Ok(res),
+                    _ => Err(format!("GET {}: wrong HTTP status '{}'", res.url(), status).into()),
                 }
             })
-            .and_then(|r| {
-                r.into_body().concat2().map_err(|e| {
-                    format!("get_manifest: failed to fetch the whole body: {}", e).into()
-                })
+            .and_then(|mut res| {
+                std::mem::replace(res.body_mut(), reqwest::async::Decoder::empty())
```

Comment:
Is this equivalent to `into_body` or is there a reason to prefer this form? I personally find it a bit obscure.

Reply:
It seems to be the only way to take ownership of the body. I've tried to use `into_body()` without success, as passing `ref` leads to an error:

```
error[E0382]: use of moved value: `res`
  --> src/v2/manifest/mod.rs:84:38
   |
81 |                 res.into_body()
   |                 --- value moved here
...
84 |                     .join(future::ok(res))
   |                                      ^^^ value used here after move
```

Reply:
Eh, the compiler is always right. I think that may work if you reverse the order of the two logical operations in the chain, ie. first extracting the header value and then collecting the body.

Reply:
I found the following to work too, would you prefer that?

```rust
            .and_then(|res| {
                future::ok(res.headers().clone()).join(
                    res.into_body()
                        .concat2()
                        .map_err(|e| Error::from(format!("{}", e))),
                )
            })
```